keps/sig-network/4410-k8s-network-interface/README.md

```diff
@@ -0,0 +1,128 @@
+
+The community may ask for more features, as we are taking a bold approach to reimagining Kubernetes networking by reducing the amount of layers involved in networking. We should prioritize feature parity with the current CNI model and then capture future work. KNI aims to be the foundational network api that is specific for Kubernetes and should make troubleshooting easier, deploying more friendly and innovate faster while reducing the need to make changes to core Kubernetes. 
+
+### Goals
+
+- Design a cool looking t-shirt
+- Provide a RPC for the Attachment and Detachment of interface[s] for a Pod
+- Provide a RPC for the Querying of Pod network information (interfaces, network namespace path, ip addresses, routes, ...)
+- Provide a RPC to prevent additional scheduling of pods if IPAM is out of IP addresses without evicting running pods
+- Provide a RPC to indicate network readiness for the node (no more CNI network configuration files in host file system)
+- Provide a RPC to provide the user the ability to query what networks are on a node
+- Consolidate K8s networking to a single layer without involving the container/oci runtimes
+- KNI should provide the RPC's required to establish feature parity with current CNI [ADD, DEL]
+- Provide documentation, examples, troubleshooting and FAQ's for KNI.
+- Decouple the Pod and Node Network setup
+- Provide garbage collection to ensure no resources created during pod setup such as Linux bridges, ebpf programs, 
+allocated IP addresses are left behind after pod deletion
+- Improve the current IP handling for pods (PodIP) to be handle multiple IP addresses and 
+a field to identify the IP address family (IPV4 vs IPV6)
+- Provide backwards compatibility for the existing CNI approach and migration a path to fully adopt KNI
+- Provide a uniform approach for network setup/teardown for both virtualized (kata) and non-virtualized (runc) 
+runtimes including kubevirt. This could eliminate the high and low level runtimes from the networking path
+- Provide a reference implementation of the KNI network runtime
+- Provide the ability to have all the dependencies packaged in the container image (no more CNI binaries in the host file system)
+..- No more downloading CNI binaries via initContainers/Mounting /etc/cni/net.d or /opt/cni/bin
```

Comment:
... Downloading a container image to run the gRPC service is better how?

Currently a cluster operator can just pre-populate these instead of relying on a daemonset, it's pretty straightforward and results in very fast startup.

Reply:
That's valid, I think - there needs to be a story for supporting pre-baked nodes.

That could be as simple as shipping the node with KNI as a system service, or shipping a node with a preloaded image.

Reply:
This doesn't make too much sense. Since calico, flannel, cilium are already running as daemonset pods. Since they would be implementing the KNI grpc service. Are you looking for a migration path here?

Reply:
There's 2 (somewhat edge case but still valid) operator questions that can be answered here.

1. I am an operator and I want to ship nodes with a specific CNI impl and not allow anyone downstream to change that, how?

2. I am an operator, and I want to set up airgapped (or at least, "fully preloaded") clusters that don't require dynamically fetching container image bytes to become functional for workload deployment.

Both of these can be solved with existing K8S primitives and patterns, with or without KNI, so I don't think KNI makes this materially more difficult.